dbms/src/Storages/KVStore/MultiRaft/RegionSerde.cpp

```diff
@@ -0,0 +1,176 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/FailPoint.h>
+#include <Storages/KVStore/Region.h>
+#include <Storages/KVStore/Utils/SerializationHelper.h>
+
+#include <memory>
+#include <utility>
+
+namespace DB
+{
+
+constexpr UInt32 Region::CURRENT_VERSION = static_cast<UInt32>(RegionPersistVersion::V2);
+
+std::pair<MaybeRegionPersistExtension, UInt32> getPersistExtensionTypeAndLength(ReadBuffer & buf)
+{
+    auto ext_type = readBinary2<MaybeRegionPersistExtension>(buf);
+    auto size = readBinary2<UInt32>(buf);
+    // Note `ext_type` may not valid in RegionPersistExtension
+    return std::make_pair(ext_type, size);
+}
+
+size_t Region::writePersistExtension(
+    UInt32 & cnt,
+    WriteBuffer & wb,
+    MaybeRegionPersistExtension ext_type,
+    const char * data,
+    UInt32 size)
+{
+    auto total_size = writeBinary2(ext_type, wb);
+    total_size += writeBinary2(size, wb);
+    wb.write(data, size);
+    total_size += size;
+    cnt++;
+    return total_size;
+}
+
+std::tuple<size_t, UInt64> Region::serialize(WriteBuffer & buf) const
+{
+    return serializeImpl(
+        Region::CURRENT_VERSION,
+        0,
+        [](UInt32 &, WriteBuffer &) { return 0; },
+        buf);
+}
+
+std::tuple<size_t, UInt64> Region::serializeImpl(
+    UInt32 binary_version,
+    UInt32 expected_extension_count,
+    std::function<size_t(UInt32 &, WriteBuffer &)> extra_handler,
+    WriteBuffer & buf) const
+{
+    size_t total_size = writeBinary2(binary_version, buf);
+    UInt64 applied_index = -1;
+
+    {
+        std::shared_lock<std::shared_mutex> lock(mutex);
+
+        // Serialize meta
+        const auto [meta_size, index] = meta.serialize(buf);
+        total_size += meta_size;
+        applied_index = index;
+
+        // Try serialize extra flags
+        if (binary_version >= 2)
+        {
+            static_assert(sizeof(eager_truncated_index) == sizeof(UInt64));
+            // The upper 31 bits are used to store the length of extensions, and the lowest bit is flag of eager gc.
+            UInt32 flags = (expected_extension_count << 1) | RegionPersistFormat::HAS_EAGER_TRUNCATE_INDEX;
```

Comment:
```
UInt32 flags = (expected_extension_count << 1) | magic_enum::enum_underlying(RegionPersistExtension::EagerTruncateIndex);
```
Maybe merge `EagerTruncateIndex = 1` into `RegionPersistExtension` and we can remove the `RegionPersistFormat`.

Reply:
This is kind of strange. Since if `EagerTruncateIndex = 1` is in `RegionPersistExtension`, other developers may wonder why we don't handle this in `deserializeImpl` and `serializeImpl`?

Reply:
OK